src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -219,6 +250,133 @@ public void finish()
+     * SSTable {@link Ref} instances. This will allow the {@link SSTableIndex} itself to be released and will fail the query.
+     */
+    private Map<Expression, NavigableSet<SSTableIndex>> referenceAndGetView(Collection<Expression> expressions)
+    {
+        SortedSet<String> indexNames = new TreeSet<>();
+        try
+        {
+            while (true)
+            {
+                List<SSTableIndex> referencedIndexes = new ArrayList<>();
+                boolean failed = false;
+
+                Map<Expression, NavigableSet<SSTableIndex>> view = getView(expressions);
+
+                for (SSTableIndex index : view.values().stream().flatMap(Collection::stream).collect(Collectors.toList()))
+                {
+                    indexNames.add(index.getIndexContext().getIndexName());
+
+                    if (index.reference())
+                    {
+                        referencedIndexes.add(index);
+                    }
+                    else
+                    {
+                        failed = true;
+                        break;
+                    }
+                }
+
+                if (failed)
+                {
+                    // TODO: This might be a good candidate for a table/index group metric in the future...
+                    referencedIndexes.forEach(QueryController::releaseQuietly);
+                }
+                else
+                {
+                    return view;
+                }
+            }
+        }
+        finally
+        {
+            Tracing.trace("Querying storage-attached indexes {}", indexNames);
+        }
+    }
+
+    private Map<Expression, NavigableSet<SSTableIndex>> getView(Collection<Expression> expressions)
+    {
+        // first let's determine the primary expression
+        Pair<Expression, NavigableSet<SSTableIndex>> primary = calculatePrimary(expressions);
+
+        Map<Expression, NavigableSet<SSTableIndex>> indexes = new HashMap<>();
+        for (Expression e : expressions)
+        {
+            // Non-index column query should only act as FILTER BY for satisfiedBy(Row) method
+            // because otherwise it likely to go through the whole index.
+            if (e.context.isNotIndexed())
+            {
+                continue;
+            }
+
+            // primary expression, we'll have to add as is
+            if (primary != null && e.equals(primary.left))
+            {
+                indexes.put(primary.left, primary.right);
+
+                continue;
+            }
+
+            View view = e.context.getView();
+
+            NavigableSet<SSTableIndex> readers = new TreeSet<>(SSTableIndex.COMPARATOR);
+            if (primary != null && primary.right.size() > 0)
+            {
+                for (SSTableIndex index : primary.right)
+                    readers.addAll(view.match(index.minKey(), index.maxKey()));
+            }
+            else
+            {
+                readers.addAll(applyScope(view.match(e)));
```

Comment:
How can there be no primary if at least one expression is against an indexed column?

Reply:
I think this logic is a bit confusing. 

Going back to the `calculatePrimary` it is possible for us to have a situation where no indexes are returned from `applyScope`? If that is the case then currently we could return a `primary` that has an empty set of indexes. As such, the above logic is nearly correct but we don't need to be checking for the `primary` being `null`.

The question is really whether we should return a `primary` in the case of no indexes and just return `null`. In which case, we have the `null` check but not the `index` size check. I'm inclined to not return a primary in this case. It would make more logical sense.

Reply:
Saying that, I'm trying to work out how we could have a query with a range that isn't satisfied by any SSTables.